loader/unknown_ext_chain_gas_aarch.S

```diff
@@ -26,6 +26,31 @@
 
 .include "gen_defines.asm"
 
+/*
+ * References:
+ *  - https://developer.arm.com/documentation/101028/0012/5--Feature-test-macros
+ *  - https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst
+ */
+#if defined(__ARM_FEATURE_BTI_DEFAULT) && __ARM_FEATURE_BTI_DEFAULT == 1
```

Comment:
So this source indicates that the `__ARM_FEATURE_BTI_DEFAULT` wont be defined in assembly sources. 

Is there a way to validate that the binaries have the correct modifications? Running the tests are good but if the assembly code never had __ARM_FEATURE_BTI_DEFAULT defined then the tests could work without indicating any issues.

https://developer.arm.com/documentation/100748/0618/Security-features-supported-in-Arm-Compiler-for-Embedded/Armv8-1-M-PACBTI-extension-mitigations-against-ROP-and-JOP-style-attacks?lang=en

Reply:
> So this source indicates that the __ARM_FEATURE_BTI_DEFAULT wont be defined in assembly sources.

I'm not fully sure if I understand your question, but I'll attempt an answer. These defines are enabled by clang and gcc within the C preprocessor, so in practice that will be assembly and language sources that utilize cpp (C pre proccesor), ie the gcc or clang front ends that do all of it. The compiler is instrumented to enable the proper landing pads when enabled, and thus only hand rolled assembly needs modifications to add the BTI instructions.

> Is there a way to validate that the binaries have the correct modifications?
You can `objdump -d` them and look at the output, as well as checking the gnu notes section using `readelf -n` on a resulting object file, executable, etc.

> Running the tests are good but if the assembly code never had __ARM_FEATURE_BTI_DEFAULT defined then the tests could work without indicating any issues.

You could have any combination of a shared library and executable BTI support and it works. The loader marks the memory mapping with `PROT_BTI` and the mmu enforces the checks. Ie calling from a BTI enabled memory region to a non-bti memory region works as well as all combinations. So their is no requirement that all memory mappings in a process have `PROT_BTI`.

Reply:
So there are two parts to my question.

1. Looking at the docs linked here, it says that the assembler doesn't define anything for you unlike the C/C++ compiler. https://developer.arm.com/documentation/100748/0618/Security-features-supported-in-Arm-Compiler-for-Embedded/Armv8-1-M-PACBTI-extension-mitigations-against-ROP-and-JOP-style-attacks?lang=en

> These attributes are output only when compiling C or C++ source. They are not output for assembly files. If you are linking with objects that are compiled with a PACBTI feature enabled, Arm recommends that you add the following code to your assembly language source files:

```c
#if !defined(__ARM_64BIT_STATE)
#ifdef __ARM_FEATURE_PAC_DEFAULT
    .eabi_attribute Tag_PAC_extension, 1
    .eabi_attribute Tag_PACRET_use, 1
#endif
#ifdef __ARM_FEATURE_BTI_DEFAULT
    .eabi_attribute Tag_BTI_extension, 1
    .eabi_attribute Tag_BTI_use, 1
#endif
#endif
```

2. Is there a tool that will tell you if PAC/BTI are properly enabled? IE some sort of test to confirm that it is working as expected. Inspecting binaries with a tool is fine, as well as any other secondary way to confirm the features are enabled. This is more of a 'peace of mind' type question than a strong requirement. I'm very unfamiliar with these security measures and am happy to defer to your expertise.

Reply:
> Looking at the docs linked here, it says that the assembler doesn't define anything for you unlike the C/C++ compiler.

Correct the assembler doesn't, but here is the important part. This assembly file is processed with gcc so the C pre-processor is invoked. That's generally the semantic difference between a .S (capital) and a .s (lowercase) suffixed assembly file is the capital S version is using the gcc/clang front end and .s lowercase is passed to the assembler directly.

For the code block that's for the EABI ABI not the Linux ABI, different ABIs.

> Is there a tool that will tell you if PAC/BTI are properly enabled? 

You can check if the ELF file declares it by running `readelf -n` on any ELF file, like the shared object:
```
readelf -n ./build/loader/libvulkan.so

Displaying notes found in: .note.gnu.property
  Owner                Data size 	Description
  GNU                  0x00000010	NT_GNU_PROPERTY_TYPE_0
      Properties: AArch64 feature: BTI, PAC
```

This will let you know if the binary supports it, and these binaries are backwards compatible and just nop on systems that don't support it. This is advantageous, say if you're a distro provider like Fedora where you want to enable the feature in a package but still support older hardware. This way the aarch64 builds just work.

As far as runtime, you need to run the existing testsuite on aarch64 machine with bti enabled:
```bash
cat /proc/cpuinfo | grep bti
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm ssbs sb paca pacg dcpodp flagm2 frint i8mm bf16 bti ecv
```

github now supports arm64 runners, but I am unaware if they support bti. If they do, it's as simple as adding the `CFLAGS/CXXFLAGS` to the build and running the testsuite.

Reply:
> This assembly file is processed with gcc so the C pre-processor is invoked. 

Well that shows my lack of experience with all of this stuff! 

> github now supports arm64 runners

I also am not sure if they are available without paying extra, so I'm not too concerned about the lack of CI testing.

You responses have eased all fears of possibly pushing bad code into the loader, so I am not going to delay merging this any further.